lazy_theta_star_p_planner/include/lazy_theta_star_p_planner/lazy_theta_star_p.hpp

```diff
@@ -0,0 +1,238 @@
+// Copyright 2020 Anshumaan Singh
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LAZY_THETA_STAR_P_PLANNER__LAZY_THETA_STAR_P_HPP_
+#define LAZY_THETA_STAR_P_PLANNER__LAZY_THETA_STAR_P_HPP_
+
+#include <cmath>
+#include <chrono>
+#include <vector>
+#include <queue>
+#include <algorithm>
+#include "rclcpp/rclcpp.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+const double INF_COST = DBL_MAX;
+const int LETHAL_COST = 252;
+
+struct coordsM
+{
+  int x, y;
+};
+
+struct coordsW
+{
+  double x, y;
+};
+
+struct pos
+{
+  int pos_id;
+  double f;
+};
+
+struct tree_node
+{
+  int x, y;
+  double g = INF_COST;
+  double h = INF_COST;
+  int parent_id;
+  bool is_in_queue = false;
+  double f = INF_COST;
+};
+
+struct comp
+{
+  bool operator()(pos & p1, pos & p2)
+  {
+    return (p1.f) > (p2.f);
+  }
+};
+
+/// TODO(ANSHU-MAN567): try using a pointer to the tree_node vector instead of directly using it
+
+namespace lazyThetaStarP
+{
+class LazyThetaStarP
+{
+public:
+  coordsM src{}, dst{};
+  nav2_costmap_2d::Costmap2D * costmap_{};
+  nav2_util::LifecycleNode::SharedPtr node_;
```

Comment:
If you pull in master or rebase, you'll see the API has changed here a little since your original PR. There's now a weak pointer not a shared pointer. You can use the weak pointer in the same way as a shared pointer but you need to call `auto node = node_.lock()` to access it. Store the weak pointer and lock and use it when necessary. 

If you need the ptr for logging / time, in configure, store a `clock` and `logger` class variables here so you only lock when you really need it

Reply:
Ah this is the pure algorithm too - why do you need the node here? Just for logging / time? If so, then especially so don't store the ptr here